File: frontend/src/api/projects.ts
import { apiRequest } from "./client";

export type ProjectStatus = "active" | "archived";
export type TaskStatus = "todo" | "in_progress" | "blocked" | "done" | "archived";
export type TaskPriority = "low" | "medium" | "high" | "urgent";

export const TASK_STATUSES: TaskStatus[] = ["todo", "in_progress", "blocked", "done"];
export const TASK_PRIORITIES: TaskPriority[] = ["low", "medium", "high", "urgent"];

export type Project = {
  id: string;
  workspace_id: string;
  created_by_id: string;
  name: string;
  description: string | null;
  slug: string;
  status: ProjectStatus;
  created_at: string;
  updated_at: string;
  archived_at: string | null;
};

export type Task = {
  id: string;
  workspace_id: string;
  project_id: string;
  parent_task_id: string | null;
  created_by_id: string;
  assignee_id: string | null;
  title: string;
  description: string | null;
  status: TaskStatus;
  priority: TaskPriority;
  due_at: string | null;
  created_at: string;
  updated_at: string;
  archived_at: string | null;
};

export type Label = {
  id: string;
  workspace_id: string;
  project_id: string;
  name: string;
  color: string;
  created_at: string;
  updated_at: string;
};

export type TaskDependency = {
  blocking_task_id: string;
  blocked_task_id: string;
  created_at: string;
};

export type Comment = {
  id: string;
  workspace_id: string;
  project_id: string;
  task_id: string;
  author_id: string;
  body: string;
  created_at: string;
  updated_at: string;
  deleted_at: string | null;
};

export type ActivityEvent = {
  id: string;
  workspace_id: string;
  project_id: string | null;
  task_id: string | null;
  actor_id: string | null;
  event_type: string;
  metadata_json: Record<string, unknown>;
  created_at: string;
};

// --- Projects ---------------------------------------------------------------

export function listProjects(token: string, workspaceId: string): Promise<Project[]> {
  return apiRequest<Project[]>(`/workspaces/${workspaceId}/projects`, { token });
}

export function createProject(
  token: string,
  workspaceId: string,
  input: { name: string; description?: string },
): Promise<Project> {
  return apiRequest<Project>(`/workspaces/${workspaceId}/projects`, {
    method: "POST",
    token,
    body: input,
  });
}

export function getProject(
  token: string,
  workspaceId: string,
  projectId: string,
): Promise<Project> {
  return apiRequest<Project>(`/workspaces/${workspaceId}/projects/${projectId}`, { token });
}

export function updateProject(
  token: string,
  workspaceId: string,
  projectId: string,
  input: { name?: string; description?: string },
): Promise<Project> {
  return apiRequest<Project>(`/workspaces/${workspaceId}/projects/${projectId}`, {
    method: "PATCH",
    token,
    body: input,
  });
}

export function archiveProject(
  token: string,
  workspaceId: string,
  projectId: string,
): Promise<void> {
  return apiRequest<void>(`/workspaces/${workspaceId}/projects/${projectId}`, {
    method: "DELETE",
    token,
  });
}

// --- Labels -------------------------------------------------------------------

export function listLabels(
  token: string,
  workspaceId: string,
  projectId: string,
): Promise<Label[]> {
  return apiRequest<Label[]>(`/workspaces/${workspaceId}/projects/${projectId}/labels`, { token });
}

export function createLabel(
  token: string,
  workspaceId: string,
  projectId: string,
  input: { name: string; color: string },
): Promise<Label> {
  return apiRequest<Label>(`/workspaces/${workspaceId}/projects/${projectId}/labels`, {
    method: "POST",
    token,
    body: input,
  });
}

// --- Tasks ----------------------------------------------------------------

export type TaskFilters = {
  status?: TaskStatus;
  assignee_id?: string;
  limit?: number;
  offset?: number;
};

export function listTasks(
  token: string,
  workspaceId: string,
  projectId: string,
  filters: TaskFilters = {},
): Promise<Task[]> {
  return apiRequest<Task[]>(`/workspaces/${workspaceId}/projects/${projectId}/tasks`, {
    token,
    query: filters,
  });
}

export function getTask(
  token: string,
  workspaceId: string,
  projectId: string,
  taskId: string,
): Promise<Task> {
  return apiRequest<Task>(`/workspaces/${workspaceId}/projects/${projectId}/tasks/${taskId}`, {
    token,
  });
}

export function createTask(
  token: string,
  workspaceId: string,
  projectId: string,
  input: {
    title: string;
    description?: string;
    assignee_id?: string | null;
    due_at?: string | null;
    priority?: TaskPriority;
    status?: TaskStatus;
    parent_task_id?: string | null;
  },
): Promise<Task> {
  return apiRequest<Task>(`/workspaces/${workspaceId}/projects/${projectId}/tasks`, {
    method: "POST",
    token,
    body: input,
  });
}

export type TaskUpdateInput = Partial<
  Pick<
    Task,
    "title" | "description" | "status" | "priority" | "assignee_id" | "due_at" | "parent_task_id"
  >
>;

export function updateTask(
  token: string,
  workspaceId: string,
  projectId: string,
  taskId: string,
  input: TaskUpdateInput,
): Promise<Task> {
  return apiRequest<Task>(`/workspaces/${workspaceId}/projects/${projectId}/tasks/${taskId}`, {
    method: "PATCH",
    token,
    body: input,
  });
}

export function archiveTask(
  token: string,
  workspaceId: string,
  projectId: string,
  taskId: string,
): Promise<void> {
  return apiRequest<void>(`/workspaces/${workspaceId}/projects/${projectId}/tasks/${taskId}`, {
    method: "DELETE",
    token,
  });
}

// --- Task labels ----------------------------------------------------------

export function listTaskLabels(
  token: string,
  workspaceId: string,
  projectId: string,
  taskId: string,
): Promise<Label[]> {
  return apiRequest<Label[]>(
    `/workspaces/${workspaceId}/projects/${projectId}/tasks/${taskId}/labels`,
    { token },
  );
}

export function addTaskLabel(
  token: string,
  workspaceId: string,
  projectId: string,
  taskId: string,
  labelId: string,
): Promise<Label[]> {
  return apiRequest<Label[]>(
    `/workspaces/${workspaceId}/projects/${projectId}/tasks/${taskId}/labels`,
    { method: "POST", token, body: { label_id: labelId } },
  );
}

export function removeTaskLabel(
  token: string,
  workspaceId: string,
  projectId: string,
  taskId: string,
  labelId: string,
): Promise<void> {
  return apiRequest<void>(
    `/workspaces/${workspaceId}/projects/${projectId}/tasks/${taskId}/labels/${labelId}`,
    { method: "DELETE", token },
  );
}

// --- Dependencies -----------------------------------------------------------

export function listDependencies(
  token: string,
  workspaceId: string,
  projectId: string,
  taskId: string,
): Promise<TaskDependency[]> {
  return apiRequest<TaskDependency[]>(
    `/workspaces/${workspaceId}/projects/${projectId}/tasks/${taskId}/dependencies`,
    { token },
  );
}

export function addDependency(
  token: string,
  workspaceId: string,
  projectId: string,
  taskId: string,
  blockingTaskId: string,
): Promise<TaskDependency> {
  return apiRequest<TaskDependency>(
    `/workspaces/${workspaceId}/projects/${projectId}/tasks/${taskId}/dependencies`,
    { method: "POST", token, body: { blocking_task_id: blockingTaskId } },
  );
}

// --- Comments ---------------------------------------------------------------

export function listComments(
  token: string,
  workspaceId: string,
  projectId: string,
  taskId: string,
): Promise<Comment[]> {
  return apiRequest<Comment[]>(
    `/workspaces/${workspaceId}/projects/${projectId}/tasks/${taskId}/comments`,
    { token },
  );
}

export function createComment(
  token: string,
  workspaceId: string,
  projectId: string,
  taskId: string,
  body: string,
): Promise<Comment> {
  return apiRequest<Comment>(
    `/workspaces/${workspaceId}/projects/${projectId}/tasks/${taskId}/comments`,
    { method: "POST", token, body: { body } },
  );
}

export function updateComment(
  token: string,
  workspaceId: string,
  projectId: string,
  taskId: string,
  commentId: string,
  body: string,
): Promise<Comment> {
  return apiRequest<Comment>(
    `/workspaces/${workspaceId}/projects/${projectId}/tasks/${taskId}/comments/${commentId}`,
    { method: "PATCH", token, body: { body } },
  );
}

export function deleteComment(
  token: string,
  workspaceId: string,
  projectId: string,
  taskId: string,
  commentId: string,
): Promise<void> {
  return apiRequest<void>(
    `/workspaces/${workspaceId}/projects/${projectId}/tasks/${taskId}/comments/${commentId}`,
    { method: "DELETE", token },
  );
}

// --- Activity ---------------------------------------------------------------

export function listActivity(
  token: string,
  workspaceId: string,
  projectId: string,
  options: { taskId?: string; limit?: number } = {},
): Promise<ActivityEvent[]> {
  return apiRequest<ActivityEvent[]>(`/workspaces/${workspaceId}/projects/${projectId}/activity`, {
    token,
    query: { task_id: options.taskId, limit: options.limit },
  });
}

File: backend/app/api/projects.py
from __future__ import annotations

import uuid
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.api.dependencies import get_current_user
from app.db.session import get_db
from app.domains.auth.models import User
from app.domains.projects import service
from app.domains.projects.models import (
    ActivityEvent,
    Label,
    Project,
    Task,
    TaskComment,
    TaskDependency,
    TaskStatus,
)
from app.domains.projects.schemas import (
    ActivityEventRead,
    CommentCreateRequest,
    CommentRead,
    CommentUpdateRequest,
    DependencyCreateRequest,
    LabelCreateRequest,
    LabelRead,
    ProjectCreateRequest,
    ProjectRead,
    ProjectUpdateRequest,
    TaskCreateRequest,
    TaskDependencyRead,
    TaskLabelRequest,
    TaskRead,
    TaskUpdateRequest,
)
from app.domains.workspaces import service as workspace_service

router = APIRouter(prefix="/workspaces/{workspace_id}", tags=["projects"])


def project_response(project: Project) -> ProjectRead:
    return ProjectRead.model_validate(project)


def task_response(task: Task) -> TaskRead:
    return TaskRead.model_validate(task)


def label_response(label: Label) -> LabelRead:
    return LabelRead.model_validate(label)


def dependency_response(dependency: TaskDependency) -> TaskDependencyRead:
    return TaskDependencyRead.model_validate(dependency)


def comment_response(comment: TaskComment) -> CommentRead:
    return CommentRead.model_validate(comment)


def activity_response(event: ActivityEvent) -> ActivityEventRead:
    return ActivityEventRead.model_validate(event)


def translate_project_error(
    error: service.ProjectError | workspace_service.WorkspaceError,
) -> HTTPException:
    if isinstance(error, workspace_service.WorkspaceNotFoundError):
        return HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Resource not found")
    if isinstance(error, workspace_service.WorkspaceArchivedError):
        return HTTPException(status_code=status.HTTP_409_CONFLICT, detail="Workspace is archived")
    if isinstance(error, service.ProjectNotFoundError | service.TaskNotFoundError):
        return HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Resource not found")
    if isinstance(error, service.LabelNotFoundError | service.CommentNotFoundError):
        return HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Resource not found")
    if isinstance(error, service.ProjectArchivedError):
        return HTTPException(status_code=status.HTTP_409_CONFLICT, detail="Project is archived")
    if isinstance(error, service.DuplicateProjectError):
        return HTTPException(status_code=status.HTTP_409_CONFLICT, detail="Project already exists")
    if isinstance(error, service.InvalidAssignmentError):
        return HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail="Invalid assignee"
        )
    if isinstance(error, service.InvalidParentTaskError):
        return HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail="Invalid parent task"
        )
    if isinstance(error, service.InvalidDependencyError):
        return HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail="Invalid dependency"
        )
    if isinstance(error, service.DuplicateDependencyError):
        return HTTPException(
            status_code=status.HTTP_409_CONFLICT, detail="Dependency already exists"
        )
    if isinstance(error, service.DuplicateLabelError):
        return HTTPException(status_code=status.HTTP_409_CONFLICT, detail="Label already exists")
    return HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Not authorized")


@router.post("/projects", response_model=ProjectRead, status_code=status.HTTP_201_CREATED)
def create_project(
    workspace_id: uuid.UUID,
    payload: ProjectCreateRequest,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
) -> ProjectRead:
    try:
        project = service.create_project(
            db,
            workspace_id,
            current_user,
            name=payload.name,
            description=payload.description,
        )
    except (service.ProjectError, workspace_service.WorkspaceError) as error:
        raise translate_project_error(error) from None
    return project_response(project)


@router.get("/projects", response_model=list[ProjectRead])
def list_projects(
    workspace_id: uuid.UUID,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
) -> list[ProjectRead]:
    try:
        return [
            project_response(project)
            for project in service.list_projects(db, workspace_id, current_user)
        ]
    except (service.ProjectError, workspace_service.WorkspaceError) as error:
        raise translate_project_error(error) from None


@router.get("/projects/{project_id}", response_model=ProjectRead)
def get_project(
    workspace_id: uuid.UUID,
    project_id: uuid.UUID,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
) -> ProjectRead:
    try:
        return project_response(
            service.get_project_for_user(db, workspace_id, project_id, current_user)
        )
    except (service.ProjectError, workspace_service.WorkspaceError) as error:
        raise translate_project_error(error) from None


@router.patch("/projects/{project_id}", response_model=ProjectRead)
def update_project(
    workspace_id: uuid.UUID,
    project_id: uuid.UUID,
    payload: ProjectUpdateRequest,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
) -> ProjectRead:
    try:
        project = service.update_project(
            db,
            workspace_id,
            project_id,
            current_user,
            name=payload.name,
            description=payload.description,
        )
    except (service.ProjectError, workspace_service.WorkspaceError) as error:
        raise translate_project_error(error) from None
    return project_response(project)


@router.delete("/projects/{project_id}", status_code=status.HTTP_204_NO_CONTENT)
def archive_project(
    workspace_id: uuid.UUID,
    project_id: uuid.UUID,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
) -> None:
    try:
        service.archive_project(db, workspace_id, project_id, current_user)
    except (service.ProjectError, workspace_service.WorkspaceError) as error:
        raise translate_project_error(error) from None


@router.post(
    "/projects/{project_id}/labels", response_model=LabelRead, status_code=status.HTTP_201_CREATED
)
def create_label(
    workspace_id: uuid.UUID,
    project_id: uuid.UUID,
    payload: LabelCreateRequest,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
) -> LabelRead:
    try:
        return label_response(
            service.create_label(
                db,
                workspace_id,
                project_id,
                current_user,
                name=payload.name,
                color=payload.color,
            )
        )
    except (service.ProjectError, workspace_service.WorkspaceError) as error:
        raise translate_project_error(error) from None


@router.get("/projects/{project_id}/labels", response_model=list[LabelRead])
def list_labels(
    workspace_id: uuid.UUID,
    project_id: uuid.UUID,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
) -> list[LabelRead]:
    try:
        return [
            label_response(label)
            for label in service.list_labels(db, workspace_id, project_id, current_user)
        ]
    except (service.ProjectError, workspace_service.WorkspaceError) as error:
        raise translate_project_error(error) from None


@router.post(
    "/projects/{project_id}/tasks", response_model=TaskRead, status_code=status.HTTP_201_CREATED
)
def create_task(
    workspace_id: uuid.UUID,
    project_id: uuid.UUID,
    payload: TaskCreateRequest,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
) -> TaskRead:
    try:
        task = service.create_task(
            db,
            workspace_id,
            project_id,
            current_user,
            title=payload.title,
            description=payload.description,
            status=payload.status,
            priority=payload.priority,
            assignee_id=payload.assignee_id,
            due_at=payload.due_at,
            parent_task_id=payload.parent_task_id,
        )
    except (service.ProjectError, workspace_service.WorkspaceError) as error:
        raise translate_project_error(error) from None
    return task_response(task)


@router.get("/projects/{project_id}/tasks", response_model=list[TaskRead])
def list_tasks(
    workspace_id: uuid.UUID,
    project_id: uuid.UUID,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
    task_status: Annotated[TaskStatus | None, Query(alias="status")] = None,
    assignee_id: uuid.UUID | None = None,
    limit: Annotated[int, Query(ge=1, le=100)] = 50,
    offset: Annotated[int, Query(ge=0)] = 0,
) -> list[TaskRead]:
    try:
        tasks = service.list_tasks(
            db,
            workspace_id,
            project_id,
            current_user,
            status=task_status,
            assignee_id=assignee_id,
            limit=limit,
            offset=offset,
        )
    except (service.ProjectError, workspace_service.WorkspaceError) as error:
        raise translate_project_error(error) from None
    return [task_response(task) for task in tasks]


@router.get("/projects/{project_id}/tasks/{task_id}", response_model=TaskRead)
def get_task(
    workspace_id: uuid.UUID,
    project_id: uuid.UUID,
    task_id: uuid.UUID,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
) -> TaskRead:
    try:
        service.require_read_membership(db, workspace_id, current_user)
        return task_response(service.get_task(db, workspace_id, project_id, task_id))
    except (service.ProjectError, workspace_service.WorkspaceError) as error:
        raise translate_project_error(error) from None


@router.patch("/projects/{project_id}/tasks/{task_id}", response_model=TaskRead)
def update_task(
    workspace_id: uuid.UUID,
    project_id: uuid.UUID,
    task_id: uuid.UUID,
    payload: TaskUpdateRequest,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
) -> TaskRead:
    try:
        return task_response(
            service.update_task(
                db,
                workspace_id,
                project_id,
                task_id,
                current_user,
                **payload.model_dump(exclude_unset=True),
            )
        )
    except (service.ProjectError, workspace_service.WorkspaceError) as error:
        raise translate_project_error(error) from None


@router.delete("/projects/{project_id}/tasks/{task_id}", status_code=status.HTTP_204_NO_CONTENT)
def archive_task(
    workspace_id: uuid.UUID,
    project_id: uuid.UUID,
    task_id: uuid.UUID,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
) -> None:
    try:
        service.archive_task(db, workspace_id, project_id, task_id, current_user)
    except (service.ProjectError, workspace_service.WorkspaceError) as error:
        raise translate_project_error(error) from None


@router.post(
    "/projects/{project_id}/tasks/{task_id}/dependencies",
    response_model=TaskDependencyRead,
    status_code=status.HTTP_201_CREATED,
)
def add_dependency(
    workspace_id: uuid.UUID,
    project_id: uuid.UUID,
    task_id: uuid.UUID,
    payload: DependencyCreateRequest,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
) -> TaskDependencyRead:
    try:
        dependency = service.add_dependency(
            db,
            workspace_id,
            project_id,
            task_id,
            current_user,
            payload.blocking_task_id,
        )
    except (service.ProjectError, workspace_service.WorkspaceError) as error:
        raise translate_project_error(error) from None
    return dependency_response(dependency)


@router.get(
    "/projects/{project_id}/tasks/{task_id}/dependencies", response_model=list[TaskDependencyRead]
)
def list_dependencies(
    workspace_id: uuid.UUID,
    project_id: uuid.UUID,
    task_id: uuid.UUID,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
) -> list[TaskDependencyRead]:
    try:
        dependencies = service.list_dependencies(
            db, workspace_id, project_id, task_id, current_user
        )
    except (service.ProjectError, workspace_service.WorkspaceError) as error:
        raise translate_project_error(error) from None
    return [dependency_response(dependency) for dependency in dependencies]


@router.post("/projects/{project_id}/tasks/{task_id}/labels", response_model=list[LabelRead])
def add_task_label(
    workspace_id: uuid.UUID,
    project_id: uuid.UUID,
    task_id: uuid.UUID,
    payload: TaskLabelRequest,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
) -> list[LabelRead]:
    try:
        labels = service.add_label_to_task(
            db, workspace_id, project_id, task_id, current_user, payload.label_id
        )
    except (service.ProjectError, workspace_service.WorkspaceError) as error:
        raise translate_project_error(error) from None
    return [label_response(label) for label in labels]


@router.get("/projects/{project_id}/tasks/{task_id}/labels", response_model=list[LabelRead])
def list_task_labels(
    workspace_id: uuid.UUID,
    project_id: uuid.UUID,
    task_id: uuid.UUID,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
) -> list[LabelRead]:
    try:
        labels = service.list_task_labels(db, workspace_id, project_id, task_id, current_user)
    except (service.ProjectError, workspace_service.WorkspaceError) as error:
        raise translate_project_error(error) from None
    return [label_response(label) for label in labels]


@router.delete(
    "/projects/{project_id}/tasks/{task_id}/labels/{label_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
def remove_task_label(
    workspace_id: uuid.UUID,
    project_id: uuid.UUID,
    task_id: uuid.UUID,
    label_id: uuid.UUID,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
) -> None:
    try:
        service.remove_label_from_task(
            db, workspace_id, project_id, task_id, current_user, label_id
        )
    except (service.ProjectError, workspace_service.WorkspaceError) as error:
        raise translate_project_error(error) from None


@router.post(
    "/projects/{project_id}/tasks/{task_id}/comments",
    response_model=CommentRead,
    status_code=status.HTTP_201_CREATED,
)
def create_comment(
    workspace_id: uuid.UUID,
    project_id: uuid.UUID,
    task_id: uuid.UUID,
    payload: CommentCreateRequest,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
) -> CommentRead:
    try:
        comment = service.create_comment(
            db, workspace_id, project_id, task_id, current_user, payload.body
        )
    except (service.ProjectError, workspace_service.WorkspaceError) as error:
        raise translate_project_error(error) from None
    return comment_response(comment)


@router.get("/projects/{project_id}/tasks/{task_id}/comments", response_model=list[CommentRead])
def list_comments(
    workspace_id: uuid.UUID,
    project_id: uuid.UUID,
    task_id: uuid.UUID,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
) -> list[CommentRead]:
    try:
        comments = service.list_comments(db, workspace_id, project_id, task_id, current_user)
    except (service.ProjectError, workspace_service.WorkspaceError) as error:
        raise translate_project_error(error) from None
    return [comment_response(comment) for comment in comments]


@router.patch(
    "/projects/{project_id}/tasks/{task_id}/comments/{comment_id}", response_model=CommentRead
)
def update_comment(
    workspace_id: uuid.UUID,
    project_id: uuid.UUID,
    task_id: uuid.UUID,
    comment_id: uuid.UUID,
    payload: CommentUpdateRequest,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
) -> CommentRead:
    try:
        comment = service.update_comment(
            db, workspace_id, project_id, task_id, comment_id, current_user, payload.body
        )
    except (service.ProjectError, workspace_service.WorkspaceError) as error:
        raise translate_project_error(error) from None
    return comment_response(comment)


@router.delete(
    "/projects/{project_id}/tasks/{task_id}/comments/{comment_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
def delete_comment(
    workspace_id: uuid.UUID,
    project_id: uuid.UUID,
    task_id: uuid.UUID,
    comment_id: uuid.UUID,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
) -> None:
    try:
        service.delete_comment(db, workspace_id, project_id, task_id, comment_id, current_user)
    except (service.ProjectError, workspace_service.WorkspaceError) as error:
        raise translate_project_error(error) from None


@router.get("/projects/{project_id}/activity", response_model=list[ActivityEventRead])
def list_activity(
    workspace_id: uuid.UUID,
    project_id: uuid.UUID,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
    task_id: uuid.UUID | None = None,
    limit: Annotated[int, Query(ge=1, le=100)] = 50,
    offset: Annotated[int, Query(ge=0)] = 0,
) -> list[ActivityEventRead]:
    try:
        events = service.list_activity(
            db,
            workspace_id,
            project_id,
            current_user,
            task_id=task_id,
            limit=limit,
            offset=offset,
        )
    except (service.ProjectError, workspace_service.WorkspaceError) as error:
        raise translate_project_error(error) from None
    return [activity_response(event) for event in events]
